File: programmers/N으로 표현.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int answer =9;

void dfs(int n, int number, int idx, int sum){
    if(idx>8) return;
    
    if(sum == number) {
        answer = min(answer, idx);
    }
    
    int tmp = 0;
    for(int i=0; i<8; i++){
        tmp = tmp*10 + n;
        dfs(n, number, idx+i+1, sum+tmp);
        dfs(n, number, idx+i+1, sum-tmp);
        dfs(n, number, idx+i+1, sum*tmp);
        dfs(n, number, idx+i+1, sum/tmp);
    }
}

int solution(int N, int number) {
    dfs(N, number, 0, 0);
    if(answer >8) return -1;
    else return answer;
}

File: programmers/N으로 표현.java
class Solution {
    
    public int answer;
    public int resultNum;
  
    void dfs(int N, int num, int cnt ,int acc){
        int nn=N;
        
        if(cnt>8){
            return;
        }
        
        if(num==acc){
            if(answer>cnt){
                answer=cnt;
            }
            return;
        }
        for(int i=1; i<9-cnt; i++ ){
            dfs(N,num,cnt+i,acc+nn);
            dfs(N,num,cnt+i,acc/nn);
            dfs(N,num,cnt+i,acc*nn);
            dfs(N,num,cnt+i,acc-nn);
            nn=nn*10+N;
        }
        
    }
    
    
    public int solution(int N, int number) {
        answer = 9;
        dfs(N,number,0,0);
        if(answer==9){
            return -1;
        }
        
        return answer;
    }
}